ophyd/signal.py

```diff
@@ -810,7 +921,116 @@ def limits(self):
+            connection_timeout = float(connection_timeout)
+
+        # Retry connecting and reading until either our entire time budget
+        # (timeout) is exhausted or our number of read_retries is reached,
+        # whichever happens first.
+        err = None
+        for attempt in range(attempts):
+            try:
+                try:
+                    self.wait_for_connection(timeout=connection_timeout)
+                except TimeoutError:
+                    raise ConnectionTimeoutError(
+                        f"Failed to connect to {self._read_pvname} "
+                        f"within {connection_timeout:.2} sec")
+                # Pyepics returns None when a read request times out.
+                # Raise a TimeoutError on its behalf.
+                info = pv.get_with_metadata(as_string=as_string,
+                                            form=form,
+                                            timeout=read_timeout)
+
+                if info is None:
+                    raise ReadTimeoutError(
+                        f"Failed to read {self._read_pvname} "
+                        f"within {read_timeout:.2} sec")
+            except TimeoutError as err_:
+                err = err_
+                # This TimeoutError could be due to connection timeout or read
+                # request timeout.
+                if deadline is not None and time.monotonic() > deadline:
+                    # We are out of time. Give up.
+                    # Notice that both errors will be shown in the traceback:
+                    # the immediate cause of this failure and a more general
+                    # message specifying how long we tried.
+                    raise TimeoutError(
+                        f"Failed to read {self._read_pvname} in "
+                        f"{timeout:.2} sec. {attempt + 1} attempt(s) were made. "
+                        f"Giving up because timeout has been reached.") from err
+                elif attempt + 1 < attempts:
+                    # We are about to continue through the retry loop.
+                    if isinstance(err, ConnectionTimeoutError):
+                        self.log.warning(
+                            "Attempt %d/%d to connect to PV %s timed out "
+                            "in %.2f sec. Retrying....",
+                            attempt + 1,
+                            attempts,
+                            self._read_pvname,
+                            connection_timeout)
+                    elif isinstance(err, ReadTimeoutError):
+                        self.log.warning(
+                            "Attempt %d/%d to read PV %s timed out "
+                            "in %.2f sec. Retrying....",
+                            attempt + 1,
+                            attempts,
+                            self._read_pvname,
+                            read_timeout)
+                    else:
+                        assert False, "Unexpected error type"
```

Comment:
This will be triggered if `pv.get_with_metadata` raises it's own `TimeoutError` rather than returning None.